Restructure CSS color parsing to be more declarative

weinig opened the pull request:
#### 9bcbb8f9e5ddd017b61405c4b528aeb0bceb2b70
<pre>
Restructure CSS color parsing to be more declarative
<a href="https://bugs.webkit.org/show_bug.cgi?id=273372">https://bugs.webkit.org/show_bug.cgi?id=273372</a>

Reviewed by Tim Nguyen.

Restructures CSS color parsing and normalization to use be driven
by a set of constexpr descriptors of the various different color
function types. This reduces redundancy, but the main goal is to
make late resolution of relative colors easier, needed for currentColor
support, easier in a forthcoming change.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Added CSSColorDescriptors.h.

* Source/WebCore/css/color/CSSColorDescriptors.h: Added.
(WebCore::RGBFunctionLegacy):
(WebCore::RGBFunctionModernAbsolute):
(WebCore::RGBFunctionModernRelative):
(WebCore::HSLFunctionLegacy):
(WebCore::HSLFunctionModern):
(WebCore::HWBFunction):
(WebCore::LabFunction):
(WebCore::LCHFunction):
(WebCore::OKLabFunction):
(WebCore::OKLCHFunction):
(WebCore::ColorRGBFunction):
(WebCore::ColorXYZFunction):
    - Added descriptors for all the CSS color functions that directly
      describe a color. The descriptors contain enough information to
      parse and normalize the the absolute and relative forms.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Angle.h:
(WebCore::CSSPropertyParserHelpers::ConsumerLookup&lt;AngleOrNumberOrNoneRaw&gt;::operator()):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Meta.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Number.h:
(WebCore::CSSPropertyParserHelpers::ConsumerLookup&lt;NumberRaw&gt;::operator()):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Percent.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Percent.h:
(WebCore::CSSPropertyParserHelpers::consumeNumberOrPercentRaw):
(WebCore::CSSPropertyParserHelpers::consumeNumberOrPercentOrNoneRaw):
(WebCore::CSSPropertyParserHelpers::consumePercentOrNoneRaw):
(WebCore::CSSPropertyParserHelpers::ConsumerLookup&lt;PercentRaw&gt;::operator()):
(WebCore::CSSPropertyParserHelpers::ConsumerLookup&lt;NumberOrPercentRaw&gt;::operator()):
(WebCore::CSSPropertyParserHelpers::ConsumerLookup&lt;NumberOrPercentOrNoneRaw&gt;::operator()):
    - Add support for invoking a consumer from a generic context based on
      the result type. For instance, if you want to consume a NumberOrPercentRaw,
      a caller now use `ConsumerLookup&lt;NumberOrPercentRaw&gt;()(args)` to call
      `consumeNumberOrPercentRaw(args)`.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
(WebCore::CSSPropertyParserHelpers::outsideSRGBGamut):
(WebCore::CSSPropertyParserHelpers::convertAbsoluteFunctionToColor):
(WebCore::CSSPropertyParserHelpers::convertRelativeFunctionToColor):
(WebCore::CSSPropertyParserHelpers::normalizeComponent):
(WebCore::CSSPropertyParserHelpers::normalizeAbsoluteComponents):
(WebCore::CSSPropertyParserHelpers::normalizeRelativeComponents):
(WebCore::CSSPropertyParserHelpers::consumeComponent):
(WebCore::CSSPropertyParserHelpers::consumeAlphaDelimiter):
(WebCore::CSSPropertyParserHelpers::consumeAbsoluteComponents):
(WebCore::CSSPropertyParserHelpers::consumeRelativeComponents):
(WebCore::CSSPropertyParserHelpers::consumeAndNormalizeAbsoluteComponents):
(WebCore::CSSPropertyParserHelpers::consumeAndNormalizeRelativeComponents):
(WebCore::CSSPropertyParserHelpers::parseGenericAbsoluteFunctionParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseGenericRelativeFunctionParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseGenericFunctionParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseRGBFunctionParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseHSLFunctionParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseColorFunctionParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseColorFunctionRaw):
(WebCore::CSSPropertyParserHelpers::parseColorFunction):
(WebCore::CSSPropertyParserHelpers::normalizeLightnessPercent): Deleted.
(WebCore::CSSPropertyParserHelpers::normalizeABPercent): Deleted.
(WebCore::CSSPropertyParserHelpers::normalizeChromaPercent): Deleted.
(WebCore::CSSPropertyParserHelpers::normalizeXYZPercent): Deleted.
(WebCore::CSSPropertyParserHelpers::normalizeRGBPercent): Deleted.
(WebCore::CSSPropertyParserHelpers::normalizeAlphaPercent): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeOriginColorRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeRGBOrHSLLegacyOptionalAlphaRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeOptionalAlphaRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeOptionalAlphaRawAllowingSymbolTableIdent): Deleted.
(WebCore::CSSPropertyParserHelpers::normalizeRGBFunctionComponent): Deleted.
(WebCore::CSSPropertyParserHelpers::normalizeRGBFunctionAlpha): Deleted.
(WebCore::CSSPropertyParserHelpers::normalizeRGBFunctionComponents): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRelativeRGBParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeRGBParametersLegacyRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeRGBParametersModernRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeRGBParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRGBParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::colorByResolvingHSLComponentsModern): Deleted.
(WebCore::CSSPropertyParserHelpers::colorByResolvingHSLComponentsLegacy): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRelativeHSLParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeHSLParametersLegacyRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeHSLParametersModernRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeHSLParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseHSLParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::normalizeHWBParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseHWBParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRelativeHWBParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeHWBParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseLabParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRelativeLabParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeLabParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseLCHParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRelativeLCHParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeLCHParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseColorFunctionForRGBTypesRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRelativeColorFunctionForRGBTypes): Deleted.
(WebCore::CSSPropertyParserHelpers::parseColorFunctionForXYZTypesRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRelativeColorFunctionForXYZTypes): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeColorFunctionParameters): Deleted.
    - Merge most of the color function parsing and normalization into a
      shared set of &quot;generic&quot; parsers and normalizers. As the parsing is
      not entirely regular among all the types, we still have some code
      to start the parsing for the following cases:

      - `rgb()/rgba()` have their own entry point `parseRGBFunctionParametersRaw`
        to allow disambiguation between the modern and legacy syntaxes.
      - `hsl()/hsla()` have their own entry point `parseHSLFunctionParametersRaw`
        to allow disambiguation between the modern and legacy syntaxes.
        Sharing this with the `rgb()/rgba()` parser would be nice, but the
        `rgb()/rgba()` has an additional requirement that all the parameters
        have the same type when using the legacy form, and any sharing ended
        up being more code and more complicated.
      - `color()` (both rgb and xyz) have their own entry point `parseColorFunctionParametersRaw`
        to allow for the additional color space parameter.
      - Everything else (`hwb()`, `lab()`, `lch()`, `oklab()`, `oklch()`) uses
        `parseGenericFunctionParametersRaw` directly. The others end up calling
        the generic code after their disambiguation.

Canonical link: <a href="https://commits.webkit.org/278370@main">https://commits.webkit.org/278370@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bccfaa5cca6d9141c0571f9fb831b978ecbfeec1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50320 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29613 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2617 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53579 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1010 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52623 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35842 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/658 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41045 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52419 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27266 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43312 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22149 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/24676 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/566 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8698 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/46661 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/630 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55165 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25416 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/555 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48449 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26679 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43495 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47481 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11042 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27541 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26409 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (approved):
Can you address the style bot errors?

Commit:
Restructure CSS color parsing to be more declarative
https://bugs.webkit.org/show_bug.cgi?id=273372

Reviewed by Tim Nguyen.

Restructures CSS color parsing and normalization to use be driven
by a set of constexpr descriptors of the various different color
function types. This reduces redundancy, but the main goal is to
make late resolution of relative colors easier, needed for currentColor
support, easier in a forthcoming change.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Added CSSColorDescriptors.h.

* Source/WebCore/css/color/CSSColorDescriptors.h: Added.
(WebCore::RGBFunctionLegacy):
(WebCore::RGBFunctionModernAbsolute):
(WebCore::RGBFunctionModernRelative):
(WebCore::HSLFunctionLegacy):
(WebCore::HSLFunctionModern):
(WebCore::HWBFunction):
(WebCore::LabFunction):
(WebCore::LCHFunction):
(WebCore::OKLabFunction):
(WebCore::OKLCHFunction):
(WebCore::ColorRGBFunction):
(WebCore::ColorXYZFunction):
    - Added descriptors for all the CSS color functions that directly
      describe a color. The descriptors contain enough information to
      parse and normalize the the absolute and relative forms.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Angle.h:
(WebCore::CSSPropertyParserHelpers::ConsumerLookup<AngleOrNumberOrNoneRaw>::operator()):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Meta.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Number.h:
(WebCore::CSSPropertyParserHelpers::ConsumerLookup<NumberRaw>::operator()):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Percent.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Percent.h:
(WebCore::CSSPropertyParserHelpers::consumeNumberOrPercentRaw):
(WebCore::CSSPropertyParserHelpers::consumeNumberOrPercentOrNoneRaw):
(WebCore::CSSPropertyParserHelpers::consumePercentOrNoneRaw):
(WebCore::CSSPropertyParserHelpers::ConsumerLookup<PercentRaw>::operator()):
(WebCore::CSSPropertyParserHelpers::ConsumerLookup<NumberOrPercentRaw>::operator()):
(WebCore::CSSPropertyParserHelpers::ConsumerLookup<NumberOrPercentOrNoneRaw>::operator()):
    - Add support for invoking a consumer from a generic context based on
      the result type. For instance, if you want to consume a NumberOrPercentRaw,
      a caller now use `ConsumerLookup<NumberOrPercentRaw>()(args)` to call
      `consumeNumberOrPercentRaw(args)`.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
(WebCore::CSSPropertyParserHelpers::outsideSRGBGamut):
(WebCore::CSSPropertyParserHelpers::convertAbsoluteFunctionToColor):
(WebCore::CSSPropertyParserHelpers::convertRelativeFunctionToColor):
(WebCore::CSSPropertyParserHelpers::normalizeComponent):
(WebCore::CSSPropertyParserHelpers::normalizeAbsoluteComponents):
(WebCore::CSSPropertyParserHelpers::normalizeRelativeComponents):
(WebCore::CSSPropertyParserHelpers::consumeComponent):
(WebCore::CSSPropertyParserHelpers::consumeAlphaDelimiter):
(WebCore::CSSPropertyParserHelpers::consumeAbsoluteComponents):
(WebCore::CSSPropertyParserHelpers::consumeRelativeComponents):
(WebCore::CSSPropertyParserHelpers::consumeAndNormalizeAbsoluteComponents):
(WebCore::CSSPropertyParserHelpers::consumeAndNormalizeRelativeComponents):
(WebCore::CSSPropertyParserHelpers::parseGenericAbsoluteFunctionParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseGenericRelativeFunctionParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseGenericFunctionParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseRGBFunctionParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseHSLFunctionParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseColorFunctionParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseColorFunctionRaw):
(WebCore::CSSPropertyParserHelpers::parseColorFunction):
(WebCore::CSSPropertyParserHelpers::normalizeLightnessPercent): Deleted.
(WebCore::CSSPropertyParserHelpers::normalizeABPercent): Deleted.
(WebCore::CSSPropertyParserHelpers::normalizeChromaPercent): Deleted.
(WebCore::CSSPropertyParserHelpers::normalizeXYZPercent): Deleted.
(WebCore::CSSPropertyParserHelpers::normalizeRGBPercent): Deleted.
(WebCore::CSSPropertyParserHelpers::normalizeAlphaPercent): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeOriginColorRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeRGBOrHSLLegacyOptionalAlphaRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeOptionalAlphaRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::consumeOptionalAlphaRawAllowingSymbolTableIdent): Deleted.
(WebCore::CSSPropertyParserHelpers::normalizeRGBFunctionComponent): Deleted.
(WebCore::CSSPropertyParserHelpers::normalizeRGBFunctionAlpha): Deleted.
(WebCore::CSSPropertyParserHelpers::normalizeRGBFunctionComponents): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRelativeRGBParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeRGBParametersLegacyRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeRGBParametersModernRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeRGBParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRGBParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::colorByResolvingHSLComponentsModern): Deleted.
(WebCore::CSSPropertyParserHelpers::colorByResolvingHSLComponentsLegacy): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRelativeHSLParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeHSLParametersLegacyRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeHSLParametersModernRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeHSLParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseHSLParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::normalizeHWBParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseHWBParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRelativeHWBParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeHWBParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseLabParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRelativeLabParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeLabParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseLCHParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRelativeLCHParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeLCHParametersRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseColorFunctionForRGBTypesRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRelativeColorFunctionForRGBTypes): Deleted.
(WebCore::CSSPropertyParserHelpers::parseColorFunctionForXYZTypesRaw): Deleted.
(WebCore::CSSPropertyParserHelpers::parseRelativeColorFunctionForXYZTypes): Deleted.
(WebCore::CSSPropertyParserHelpers::parseNonRelativeColorFunctionParameters): Deleted.
    - Merge most of the color function parsing and normalization into a
      shared set of "generic" parsers and normalizers. As the parsing is
      not entirely regular among all the types, we still have some code
      to start the parsing for the following cases:

      - `rgb()/rgba()` have their own entry point `parseRGBFunctionParametersRaw`
        to allow disambiguation between the modern and legacy syntaxes.
      - `hsl()/hsla()` have their own entry point `parseHSLFunctionParametersRaw`
        to allow disambiguation between the modern and legacy syntaxes.
        Sharing this with the `rgb()/rgba()` parser would be nice, but the
        `rgb()/rgba()` has an additional requirement that all the parameters
        have the same type when using the legacy form, and any sharing ended
        up being more code and more complicated.
      - `color()` (both rgb and xyz) have their own entry point `parseColorFunctionParametersRaw`
        to allow for the additional color space parameter.
      - Everything else (`hwb()`, `lab()`, `lch()`, `oklab()`, `oklch()`) uses
        `parseGenericFunctionParametersRaw` directly. The others end up calling
        the generic code after their disambiguation.

Canonical link: https://commits.webkit.org/278370@main
